### PR TITLE
Include private members in API hash of traits.

### PR DESCRIPTION
**Pull request by Duhemm**
During compilation, scalac generates getters and setters for the fields
of traits, regardless of their access modifiers. Therefore, when a
field of a trait is modified, all its implementors must be recompiled to
take these changes into account.

Private fields of traits were not included in the API hash of traits,
and their implementors were thus not recompiled when modified.

This commit changes the way the API hash is computed for traits only, so
that the generated hash includes the private members of traits.

Fixes sbt/sbt#2155

### CODE REVIEW

File: compile/api/src/main/scala/xsbt/api/HashAPI.scala [hunk 1]
```diff
@@ -343,14 +345,14 @@ final class HashAPI(includePrivate: Boolean, includeParamNames: Boolean, include
     hashType(a.baseType)
     hashAnnotations(a.annotations)
   }
-  final def hashStructure(structure: Structure, includeDefinitions: Boolean) =
-    visit(visitedStructures, structure)(structure => hashStructure0(structure, includeDefinitions))
-  def hashStructure0(structure: Structure, includeDefinitions: Boolean): Unit = {
+  final def hashStructure(structure: Structure, includeDefinitions: Boolean, isTrait: Boolean = false) =
+    visit(visitedStructures, structure)(structure => hashStructure0(structure, includeDefinitions, isTrait))
+  def hashStructure0(structure: Structure, includeDefinitions: Boolean, isTrait: Boolean = false): Unit = {
```
Comment (Duhemm) [inline compile/api/src/main/scala/xsbt/api/HashAPI.scala:356]: Does that count as a breakage of the binary compatibility?

**Comment by typesafe-tools** [comment]
Can one of the admins verify this patch?

File: compile/api/src/main/scala/xsbt/api/HashAPI.scala (hunk 1 above)
Comment (eed3si9n) [inline compile/api/src/main/scala/xsbt/api/HashAPI.scala:356]: Let's try the overload, and deprecate the old one.

Comment (Duhemm) [inline compile/api/src/main/scala/xsbt/api/HashAPI.scala:356]: @eed3si9n Done

**Comment by eed3si9n** [comment]
LGTM

**Comment by dwijnand** [comment]
Nice!

**Comment by indrajitr** [comment]
:star:

**Pull request merged**